kore/src/KoreJson.hs

```diff
@@ -0,0 +1,603 @@
+import Kore.Parser (embedParsedPattern)
+import Kore.Syntax qualified as Kore
+import Kore.Syntax.PatternF (PatternF (..))
+import Kore.Syntax.Variable (
+    ElementVariableName (..),
+    SetVariableName (..),
+    SomeVariableName (..),
+    Variable (..),
+    VariableName (..),
+ )
+import Prelude.Kore hiding (Left, Right, pred)
+import Prelude.Kore qualified as Prelude (Either (..))
+
+{- | Json representation of Kore patterns as a Haskell type.
+ Modeled after kore-syntax.md, merging some of the ML pattern
+ productions. Cursorily checked to be consistent with Parser.y
+-}
+data KorePattern
+    = -- variable pattern
+
+      -- | element variable with sort
+      KJEVar
+        { name :: Id
+        , sort :: Sort
+        }
+    | -- | set variable with sort
+      KJSVar
+        { name :: Id -- must start by '@'
+        , sort :: Sort
+        }
+    | -- application pattern
+
+      -- | application pattern (symbol, [arg] )
+      KJApp
+        { name :: Id -- may start by a '\\'
+        , sorts :: [Sort]
+        , args :: [KorePattern]
+        }
+    | -- | string literal
+      KJString Text
+    | -- matching logic pattern
+
+      -- | Connective (top, bottom, not, and, or, implies, iff)
+      KJConnective
+        { conn :: Connective
+        , sort :: Sort
+        , args :: [KorePattern] -- arg count checked in conversion to TermLike, not here
+        }
+    | -- | Quantifiers: forall, exists
+      KJQuantifier
+        { quant :: Quant
+        , sort :: Sort
+        , var :: Id
+        , varSort :: Sort
+        , arg :: KorePattern
+        }
+    | -- | mu, nu
+      KJFixpoint
+        { combinator :: Fix
+        , -- no sort
+          var :: Id
+        , varSort :: Sort
+        , arg :: KorePattern
+        }
+    | -- | ceil, floor, equals, in. NOTE these do not really fit together
+      -- nicely... the two sorts have vastly different meaning
+      KJPredicate
+        { pred :: Pred
+        , sort :: Sort
+        , sort2 :: Sort
+        , args :: [KorePattern]
+        }
+    | -- next, rewrites
+
+      -- | goes to 'dest' next
+      KJNext
+        { sort :: Sort
+        , dest :: KorePattern
+        }
+    | -- | source rewrites to dest (same sort)
+      KJRewrites
+        { sort :: Sort
+        , source :: KorePattern
+        , dest :: KorePattern
+        }
+    | -- | domain value, a string literal with a sort
+      KJDomainValue
+        { sort :: Sort
+        , value :: Text
+        }
+    | -- syntactic sugar
+
+      -- | left/right associative or-cascade
+      KJMultiOr
+        { assoc :: LeftRight
+        , sort :: Sort
+        , args :: [KorePattern]
+        }
+    | -- TODO textual parser also understands And/Implies/Iff
+
+      -- | left/right associative app pattern
+      KJMultiApp
+        { assoc :: LeftRight
+        , symbol :: Id -- may start by a '\\'
+        , sorts :: [Sort]
+        , args :: [KorePattern]
+        }
```

Comment:
I think, in theory at least, this can include any binary operation, so they could be collapsed into a single, general, category.